### ISSUE TITLE
NSIS uninstaller for windows ignores silent flag

### ISSUE THREAD

**Issue by nett00n** [freeorion/freeorion#3383]
Bug Report
==========

Environment
-----------

* **FreeOrion Version**: 0.4.10.1
* **Operating System**: Windows 10 20H2
* **Graphic card used**: GTX 1070
* **Fetched as** 
  * Binary release

Description
-----------
The NSIS package from releases page successfully installs in silent mode without any popupping window:
`\FreeOrion_v0.4.10.1_2020-09-25.39cfe10_Win32_Setup.exe /S`
but uninsaller ignores silent flag `/S` and even complete silent flag `SD` and shows confirmation window "Are you sure you want to remove FreeOrion"

Expected Result
---------------
FreeOrion removes silently

Steps to reproduce
------------------

* Use any supported Windows version
* Download latest Win32_Setup.exe release
* Open CMD or PowerShell Window
* Run installer with `/S` flag
![image](https://user-images.githubusercontent.com/1846116/113500218-d764c980-9524-11eb-8207-6b5794b6df55.png)

* Wait a few moment - package will install fully in background
* Run uninsaller with silent flag `/S` or with very silent flag `/SD`
![image](https://user-images.githubusercontent.com/1846116/113500277-5528d500-9525-11eb-9de5-33498c43a5c0.png)

* Confirmation window appers (it's not okay)
![image](https://user-images.githubusercontent.com/1846116/113500106-1e9e8a80-9524-11eb-9a5a-62115d2c7d87.png)

**Comment by geoffthemedio** [comment]
Do most other NSIS installers skip the popup when executed that way?

**Comment by nett00n** [comment]
Silent behaviour is presented in NSIS specification.
Yes, most of NSIS packges removes without confirmation with `/S` flag

**Comment by Vezzra** [comment]
I assume the issue is with this section of the NSIS script:
```
Function un.onInit
  MessageBox MB_ICONQUESTION|MB_YESNO|MB_DEFBUTTON2 "Are you sure you want to completely remove $$(^Name) and all of its components?" IDYES +2
  Abort
FunctionEnd
```
Obviously, there is no check for the silent flag. However, according to @nett00n the silent flags works fine when installing, which is a bit confusing, as there is this section handling initialization of the install process:
```
Function .onInit
  ReadRegStr $$R0 HKLM "$${PRODUCT_UNINST_KEY}" "UninstallString"
  StrCmp $$R0 "" fin_onInit

  MessageBox MB_OKCANCEL|MB_ICONEXCLAMATION \
  "$${PRODUCT_NAME} is already installed. $$\n$$\nClick `OK` to run the \
  uninstall dialog (recommended) $$\nor click `Cancel` to continue." \
  IDOK uninst_onInit
  Goto fin_onInit

  uninst_onInit:
  ClearErrors
  InitPluginsDir
  CopyFiles "$$R0" "$$pluginsdir"
  ExecWait '$$pluginsdir\Uninstall.exe _?=$$INSTDIR'

  fin_onInit:

FunctionEnd
```
As you can see, there is also a call to `MessageBox` with no check of the silent flag. Of course, that message box only gets displayed if you try to install FO on a system where there is already another version installed. @nett00n, can you verify that installing FO also disregards the silent flag if you try to install over an existing FO version?

Just to make sure that the issue here indeed is that we do not check for that flag in the script, before we try to come up with fixes.

**Comment by nett00n** [comment]
@Vezzra no, the installer with /S flag installs package completely silent. 
This behavior presented only with uninstall function

**Comment by Vezzra** [comment]
@nett00n, well, in that case this can't be an issue with our install script, but with the NSIS software. As I didn't have the latest version installed, I now did that.

Meaning, the next weekly test build installer package for Windows will be created with the latest release of NSIS. Please test if the issue persists with that installer (once it is available, usually Tuesday evening) and report back.

**Comment by geoffthemedio** [comment]
Might it be related to extra files being generated on disk after installation when the program is run?

**Comment by Vezzra** [comment]
@geoffthemedio,

> Might it be related to extra files being generated on disk after installation when the program is run?

I don't know why that would have anything to do with if popup dialogs are supressed or not, but what do I know. Meaning, I actually have no idea...

**Comment by Vezzra** [comment]
@nett00n, have you been able to test the installer of the latest weekly build?

**Comment by geoffthemedio** [comment]
Idea was that if there are extra files present that it doesn't know should be there (due to their not being in the list of files that were placed there by the installer), it might refuse to silently delete those files.

**Comment by Vezzra** [comment]
@geoffthemedio, the message box the OP is complaining about is most likely caused by this section in the NSIS installer script:
```
Function un.onInit
  MessageBox MB_ICONQUESTION|MB_YESNO|MB_DEFBUTTON2 "Are you sure you want to completely remove $$(^Name) and all of its components?" IDYES +2
  Abort
FunctionEnd
```
Meaning, this isn't a "standard"/"default" message box, but a custom one, with no checks for whatever files might exist or not exist at all. This message box is going to get displayed "on uninstall init" unconditionally, so I wouldn't expect it to care anything about extra files. Can't rule out the possibility of course.

**Comment by nett00n** [comment]
@Vezzra 
i tried with last available package FreeOrion_2021-04-27.4b17969_Test_Win32_Setup.exe and it still does not uninstalls silently

**Comment by Vezzra** [comment]
@nett00n, thanks for reporting back.

That means I need to figure out how to tell NSIS to check for the silent flag before displaying this message box... oh joy... 😉

**Comment by Vezzra** [comment]
Fixed in e956fd470cac4f470dd458d722b76ead279d4f01

**Comment by Vezzra** [comment]
@nett00n, the issue should be fixed now, and will already be in the next weekly test builds. Please test the windows installer/uninstaller once they are available, and report back if the problem persists (in that case we'll reopen this issue).
